core/src/main/java/io/ddf/content/Schema.java

```diff
@@ -1,655 +1,193 @@
-    this.initialize(tableName, this.parseColumnList(columns));
-  }
-
-  @Deprecated
-  // Require tableName at all times, even null
-  public Schema(List<Column> columns) throws DDFException {
-    this.initialize(null, columns);
-  }
-
-  public Schema(String tableName, List<Column> columns) throws DDFException {
-    this.initialize(tableName, columns);
-  }
-
-  @SuppressWarnings("unchecked")
-  public Schema(String tableName, Column[] columns) throws DDFException {
-    this.initialize(tableName, Arrays.asList(columns));
-  }
-
-  private void initialize(String tableName, List<Column> columns) throws DDFException {
-    this.setColumns(columns);
-    this.mTableName = tableName;
-  }
-
-  public static void validateSchema(Schema schema) throws DDFException {
-    if(schema != null && schema.getColumns() != null) {
-      validateColumnNames(schema.getColumnNames());
-    }
-  }
-
-  private static void validateColumnNames(List<String> names) throws DDFException {
-    Set<String> columnSet = new HashSet<String>();
-    if(names != null) {
-      for(String name: names) {
-        if(columnSet.contains(name)) {
-          throw new DDFException(String.format("Duplicated column name %s", name));
-        } else {
-          if(!Utils.isAlphaNumeric(name)) {
-            throw new DDFException(String.format("Invalid column name %s, only allow alphanumeric (uppercase and lowercase a-z, numbers 0-9) " +
-                "and dash (\"-\") and underscore (\"_\")", name));
-          }
-          columnSet.add(name);
-        }
-      }
-    } else {
-      throw new DDFException("names is null");
-    }
-  }
-
-  private List<Column> parseColumnList(String columnList) {
-    if (Strings.isNullOrEmpty(columnList))
-      return null;
-    String[] segments = columnList.split("\\s*,\\s*");
+@NotThreadSafe
+public  class Schema {
 
-    mColumns.clear();
-    for (String segment : segments) {
-      if (Strings.isNullOrEmpty(segment))
-        continue;
+  protected List<String> colNames;
+  protected List<Column> columns;
 
-      String[] parts = segment.split("\\s+");
-      if (Strings.isNullOrEmpty(parts[0])
-          || Strings.isNullOrEmpty(parts[1]))
-        continue;
-
-      mColumns.add(new Column(parts[0], parts[1]));
-    }
-
-    return mColumns;
+  public Schema() {
+    colNames = new ArrayList<>();
+    columns = new ArrayList<>();
   }
 
-  public String getTableName() {
-    return mTableName;
+  public Schema(Schema schema) {
+    this();
+    this.colNames.addAll(schema.colNames);
+    columns.addAll(schema.getColumns());
+    assert colNames.size() == columns.size();
   }
 
-  public void setTableName(String mTableName) {
-    this.mTableName = mTableName;
+  public int getNumColumn() {
+    return columns.size();
   }
 
   public List<Column> getColumns() {
-    List<Column> columns = new ArrayList<Column>();
-    for (Column column : mColumns) {
-      columns.add(column.clone());
-    }
```

Comment:
getColumns should clone the list of columns and return the cloned list, avoiding user to modify the schema by modifying the columns